src/storage/src/sink/kafka.rs

```diff
@@ -910,16 +916,39 @@ impl KafkaSinkState {
     }
 
     /// Report a SinkStatus::Stalled and then halt with the same message.
-    pub async fn halt_on_err<T>(&self, result: Result<T, impl ToString + Debug>) -> T {
+    pub async fn halt_on_err<T>(&self, result: Result<T, anyhow::Error>) -> T {
         match result {
             Ok(t) => t,
-            Err(msg) => {
-                self.update_status(SinkStatus::Stalled(msg.to_string()))
-                    .await;
+            Err(error) => {
+                let hint: Option<String> =
+                    error
+                        .downcast_ref::<RDKafkaError>()
```

Comment:
personally I like this downcast, but let me check with @bkirwi real quick!

Reply:
Seems fine! If the cast doesn't succeed we just fall back to not providing a hint, which seems very peaceful.